### PR TITLE
[Storage][Sharding] Sharded state merkle pruner.

### PR DESCRIPTION
**Pull request by grao1991**
### Test Plan
Tested in executor-benchmark.

### CODE REVIEW

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 1]
```diff
@@ -121,96 +125,176 @@ where
...
+    fn update_shard_progress(&self, shard_id: u8, progress: Version) {
+        self.shard_progresses.lock()[shard_id as usize] = progress;
+    }
+
+    // If the existing schema batch is not none, this function only adds items need to be
+    // deleted to the schema batch and the caller is responsible for committing the schema batches
+    // to the DB.
+    fn prune_state_merkle_shard(
+        &self,
+        db: &DB,
```
Comment (areshand) [inline storage/aptosdb/src/pruner/state_store/mod.rs:260]: nit: could you name this DB? too many DBs, sometime, hard to know which DB this refers to

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 2]
```diff
@@ -121,96 +125,176 @@ where
...
+        target_version: Version,
+        batch_size: usize,
+    ) -> Result<()> {
+        let _timer = OTHER_TIMERS_SECONDS
+            .with_label_values(&["state_merkle_pruner___prune_single_shard"])
+            .start_timer();
+        let shard_min_readable_version = self.get_shard_progress(shard_id);
+        if shard_min_readable_version != target_version {
+            assert_lt!(shard_min_readable_version, target_version);
+            self.update_shard_progress(shard_id, target_version);
```
Comment (areshand) [inline storage/aptosdb/src/pruner/state_store/mod.rs:207]: Ideally, do we want to update shard progress once the prune is done?

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 3]
```diff
@@ -121,96 +125,176 @@ where
...
 
-                self.save_min_readable_version(new_min_readable_version, &batch)?;
+    fn prune_top_levels(
+        &self,
+        min_readable_version: Version,
+        target_version: Version,
+    ) -> Result<Option<Version>> {
+        let batch = SchemaBatch::new();
+        let next_version = self.prune_state_merkle_shard(
+            self.state_merkle_db.metadata_db(),
```
Comment (areshand) [inline storage/aptosdb/src/pruner/state_store/mod.rs:183]: is the top level SMT stored in metadata_db?

Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/mod.rs:185]: Yes

File: storage/aptosdb/src/pruner/state_store/mod.rs (hunk 2 above)
Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/mod.rs:209]: The db progress is updated after, the in memory one is updated before, to tell the upper layer the data is not readable.

File: storage/aptosdb/src/pruner/state_store/mod.rs (hunk 1 above)
Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/mod.rs:271]: Done

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 4]
```diff
@@ -27,16 +28,25 @@ mod test;
 
 pub const STATE_MERKLE_PRUNER_NAME: &str = "state_merkle_pruner";
 
+static POOL: Lazy<rayon::ThreadPool> = Lazy::new(|| {
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/mod.rs]: Done

**Review by areshand — APPROVED** [review]

File: storage/aptosdb/src/pruner/state_store/mod.rs (hunk 2 above)
Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/mod.rs]: What I said above has changed now. Now we manage the "min_readable_version" in the pruner manager (updated before the work, and this is the value we serve to outside), and we manager the "progress" in the pruner structs (updated after the work is done).

File: storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+impl<S: StaleNodeIndexSchemaTrait> StateMerkleMetadataPruner<S>
+where
+    StaleNodeIndex: KeyCodec<S>,
+{
+    pub(in crate::pruner) fn new(metadata_db: Arc<DB>) -> Self {
+        Self {
+            metadata_db,
+            next_version: AtomicVersion::new(0),
```
Comment (msmouse) [inline storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs:35]: Man I have to say this field is useless.. looking at the code it's always been overwritten by `current_progress`.

Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs:35]: Not really, for most of the time it's always be larger than current_progress, and it's overwritten by either the second smallest version that is larger than current_progress (the smallest version >= current_progress get pruned), or the target_version (if we've already reach the end of stale index in this round)

File: storage/aptosdb/src/pruner/state_store/generics.rs [hunk 6]
```diff
@@ -12,13 +12,17 @@ pub trait StaleNodeIndexSchemaTrait: Schema<Key = StaleNodeIndex>
...
 {
-    fn tag() -> DbMetadataKey;
+    fn tag(shard_id: Option<u8>) -> DbMetadataKey;
     fn name() -> &'static str;
 }
 
 impl StaleNodeIndexSchemaTrait for StaleNodeIndexSchema {
-    fn tag() -> DbMetadataKey {
-        DbMetadataKey::StateMerklePrunerProgress
+    fn tag(shard_id: Option<u8>) -> DbMetadataKey {
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/generics.rs:20]: Done

File: storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+        self.next_version
+            // If next_version is None, meaning we've already finished all pending work.
+            .store(next_version.unwrap_or(target_version), Ordering::SeqCst);
+
+        Ok(Some(target_version_for_this_round))
+    }
+
+    pub(in crate::pruner) fn progress(&self) -> Result<Version> {
+        Ok(get_progress(&self.metadata_db, &S::tag(None))?.unwrap_or(0))
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs]: Done

File: storage/aptosdb/src/pruner/state_store/mod.rs [hunk 8]
```diff
@@ -102,17 +104,19 @@ where
 
     // used only by blanket `initialize()`, use the underlying implementation instead elsewhere.
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/pruner/state_store/mod.rs]: Removed.

File: storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs (hunk 5 above)
Comment (msmouse) [reply storage/aptosdb/src/pruner/state_store/state_merkle_metadata_pruner.rs:35]: okay.. needed a vc session to understand what you are saying, but seems legit 😞

File: storage/aptosdb/src/utils/mod.rs [hunk 9]
```diff
@@ -4,19 +4,15 @@
...
-    Ok(
-        if let Some(DbMetadataValue::Version(progress)) =
-            db.get::<DbMetadataSchema>(progress_key)?
-        {
-            Some(progress)
-        } else {
-            None
-        },
-    )
+    Ok(if let Some(v) = db.get::<DbMetadataSchema>(progress_key)? {
```
Comment (msmouse) [inline storage/aptosdb/src/utils/mod.rs:13]: use `.map()`?

**Review by msmouse — APPROVED** [review]

File: storage/aptosdb/src/utils/mod.rs (hunk 9 above)
Comment (grao1991) [reply storage/aptosdb/src/utils/mod.rs]: Done

**Pull request merged**